apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PoolRewrite/SessionPool.TargetedSessionPool.cs

```diff
@@ -0,0 +1,602 @@
+
+            /// <summary>
+            /// Thread-safe read-only access to in-flight session creation count.
+            /// </summary>
+            internal int InFlightSessionCreationCount => Interlocked.CompareExchange(ref _inFlightSessionCreationCount, 0, 0);
+
+            // Statistics maintained purely for diagnostic purposes. This lets us evaluate
+            // how effective transaction pre-warming is.
+            private long _rwTransactionRequests;
+            private long _rwTransactionRequestsPrewarmed;
+
+            internal TargetedSessionPool(SessionPool parent, DatabaseName databaseName, bool acquireSessionsImmediately) : base(parent)
+            {
+                _parent = parent;
+                _databaseName = GaxPreconditions.CheckNotNull(databaseName, nameof(databaseName));
+                _createSessionRequest = new CreateSessionRequest { DatabaseAsDatabaseName = databaseName };
+                if (acquireSessionsImmediately)
+                {
+                    StartAcquisitionTasksIfNecessary();
+                }
+            }
+
+            public async Task<PooledSession> AcquireSessionAsync(TransactionOptions transactionOptions, CancellationToken cancellationToken)
+            {
+                // We may need to start more tasks just acquire a session. We need to try this *before* counting the requested
+                // session as being active. (If we have 99 active sessions at the moment, and nothing in the pool, we want to start a task
+                // to create the session we'll end up acquiring. If we wait until we've counted this session as being active, we won't
+                // start the new task as it'll take us over the limit.)
+                StartAcquisitionTasksIfNecessary();
+
+                bool success = false;
+                try
+                {
+                    Interlocked.Increment(ref _activeSessionCount);
+                    PooledSession session = await AcquireSessionImplAsync(transactionOptions, cancellationToken).ConfigureAwait(false);
+                    success = true;
+                    return session;
+                }
+                finally
+                {
+                    if (!success)
+                    {
+                        Interlocked.Decrement(ref _activeSessionCount);
+                    }
+                }
+            }
+
+            private async Task<PooledSession> AcquireSessionImplAsync(TransactionOptions transactionOptions, CancellationToken cancellationToken)
+            {
+                var transactionMode = transactionOptions?.ModeCase ?? ModeOneofCase.None;
+                var sessionAcquisitionTask = GetSessionAcquisitionTask(transactionMode, cancellationToken);
+
+                var session = await sessionAcquisitionTask.ConfigureAwait(false);
+
+                // Now we've "taken" this session, we may want to start acquiring more to replenish the pool
+                StartAcquisitionTasksIfNecessary();
```

Comment:
Althouhg I think it doesn't hurt to leave this here, I also think it does nothing. If you do the math (knowing that you lauched the `StartAcquisitionTasks...` earlier) I think there's no chance this call will launch extra tasks:

```
// How many more requests do we need to make in order to get the pool to the minimum size, after satisfying
// pending callers, assuming that all the in-flight requests succeed, and there are no more requests?
int newRequestsToSatisfyPool = (pendingAcquisitionCount + minPoolSize) - (poolSize + inFlightRequests);
// How many more requests *can* we make without going over the maximum number of active sessions?
int maxAvailableRequests = maxActiveSessions - (activeSessions + poolSize + inFlightRequests);

int actualNewRequests = Math.Min(newRequestsToSatisfyPool, maxAvailableRequests);
```
Actually, trying to do the math, I think I found a related issue, that if fixed will remove this problem here. I'm adding a new comment on `AcquireSessionAsync`.

Reply:
At this point, if we took a session from the pool, the pool will be smaller - we want to bring it back up to size if it's now below the minimum size. There are still issues which will probably affect this call, but I think it can't just be removed or we'll end up with too few sessions in the pool.